processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/XmlElementDeclSelector.java

```diff
@@ -112,4 +119,38 @@ else if ( nameMatches.size() > 0 ) {
             return methods;
         }
     }
+
+    private Pair getMatch(SourceMethod sourceMappingMethod, final String targetPropertyName) {
+        final TypeMirror startingMirror = sourceMappingMethod.getExecutable().getReturnType();
+        TypeMirror currentMirror = startingMirror;
+        TypeElement currentElement = sourceMappingMethod.getResultType().getTypeElement();
+
+        while (currentElement != null) {
+            for (Element enclosed : currentElement.getEnclosedElements()) {
+                if (enclosed.getKind().equals( ElementKind.FIELD ) && targetPropertyName != null
+                        && enclosed.getSimpleName().contentEquals( targetPropertyName )) {
```

Comment:
I'm having problems to understand this loop. Can you add some comments here what you are trying to achieve?

The problem is that the selectors are used for both selecting factory methods and mapping methods. This case is very much geared towards finding factory methods in an object factory.

What I understand: you are looking for fields in the target types of the candidate mapping methods? What you need is (I think) the target bean, or better, an `XmlElementRef` `AnnotationMirror`. Wouldn't it be better to add those to the selection criteria when you call the whole mapper chain? Thats what the `SelectionCriteria` are for.

You mentioned somewhere in the discussion that the `targetType` should be called `targetPropertyType`. I now understand your remark better :smile: . But on this level, we're just trying to find a mapping method for mapping a `sourceType` to a `targetType`. There is no understanding/knowledge of bean properties here. Selection context should come via the `SelectionCriteria`.